compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorLowering.cpp

```diff
@@ -183,6 +183,261 @@ struct SetMulAddFMF final : OpRewritePattern<vector::MultiDimReductionOp> {
+    } else {
+      flattenedAcc = vector::BroadcastOp::create(rewriter, loc, flatAccVecType,
+                                                 op.getAcc());
+    }
+
+    constexpr int64_t chunkSize = 2;
+    Value resultFlat = buildFMAChain(rewriter, loc, lhs2D, rhs2D, flattenedAcc,
+                                     redSize, parSize, chunkSize);
+
+    // Restore result to original form.
+    Value result;
+    if (accVecType) {
+      Value reshaped = vector::ShapeCastOp::create(
+          rewriter, loc, preFlattenVecType, resultFlat);
+
+      if (!isIdentityPermutation(accPerm)) {
+        result = vector::TransposeOp::create(rewriter, loc, accVecType,
+                                             reshaped, invert(accPerm));
+      } else {
+        result = reshaped;
+      }
+
+    } else {
+      result = vector::ExtractOp::create(rewriter, loc, resultFlat, 0);
+    }
+
+    rewriter.replaceOp(op, result);
+    return success();
+  }
+
+private:
+  static bool verifyParallelDimsInMap(ArrayRef<int64_t> parallelDims,
+                                      AffineMap map) {
+    llvm::SmallSetVector<int64_t, 8> usedDims;
+    map.walkExprs([&](AffineExpr expr) {
+      if (auto dimExpr = dyn_cast<AffineDimExpr>(expr)) {
+        usedDims.insert(dimExpr.getPosition());
+      }
+    });
+
+    return llvm::all_of(parallelDims, [&](int64_t parDim) {
+      return usedDims.contains(parDim);
+    });
+  }
+
+  static SmallVector<int64_t> invert(ArrayRef<int64_t> perm) {
+    SmallVector<int64_t> inv(perm.size());
+    for (int64_t i = 0; i < perm.size(); ++i) {
+      inv[perm[i]] = i;
+    }
+    return inv;
+  }
+
+  static void getReductionAndParallelLoopDims(ArrayAttr iters,
+                                              SmallVectorImpl<int64_t> &red,
+                                              SmallVectorImpl<int64_t> &par) {
+    for (auto [idx, attr] : llvm::enumerate(iters)) {
+      if (vector::isReductionIterator(attr)) {
+        red.push_back(idx);
+      } else {
+        par.push_back(idx);
+      }
+    }
+  }
+
+  /// Constructs a permutation for vector.transpose from an affine map and a
+  /// reordered list of dimension.
+  ///
+  /// Example:
+  ///   map: (d0, d1, d2) -> (d0, d2, d1)
+  ///   iterator_types = ["parallel","parallel","reduction"]
+  //    ==> new dim order: [2, 0, 1]
+  ///
+  ///   Step 1: Build dim-to-result mapping from the map.
+  ///           dimToRes = [0, 2, 1] i.e {0: 0, 1: 2, 2: 1}
+  ///
+  ///   Step 2: Walk new dimension order in order to build permutation.
+  ///           indices[0]=2 -> dimToRes[2]=1
+  ///           indices[1]=0 -> dimToRes[0]=0
+  ///           indices[2]=1 -> dimToRes[1]=2
+  ///
+  ///   Result: perm = [1, 0, 2]
+  static SmallVector<int64_t>
+  getPermutationFromIndexingMap(AffineMap map, ArrayRef<int64_t> indices) {
+    SmallVector<int64_t> dimToRes(map.getNumDims());
+    for (int res = 0; res < map.getNumResults(); ++res) {
```

Comment:
Do not recalculate the end index: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

```suggestion
    for (int res = 0, e = map.getNumResults(); res != e ; ++res) {
```